File: 프로그래머스/2/84512. 모음 사전/모음 사전.js
function solution(word) {
    
    const alphabets = ['A', 'E', 'I', 'O', 'U'];
    
    let words = [];

    function makeWord(current) {
        if (current.length > 5) return; // 길이가 5를 초과하면 종료
        
        if (current.length > 0) words.push(current); // 단어를 추가
        
        for (let i = 0; i < alphabets.length; i++) {
            makeWord(current + alphabets[i]); // 재귀적으로 단어 생성
        }
    }

    makeWord('');

    words.sort(); // 단어들을 사전순으로 정렬
    return words.indexOf(word) + 1; // 인덱스를 반환 (1부터 시작)
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;
import java.io.*;

class Solution {
    
    static String[] alphabet = {"A", "E", "I", "O", "U"};
    static ArrayList<String> list;
    
    public int solution(String word) {
        
        list = new ArrayList<>();
        
        comb("");
        
        int answer = 0;
        
        Collections.sort(list);
        
        for(int i = 0; i < list.size(); i++) {
            if(list.get(i).equals(word)) {
                answer = i + 1;
                break;
            }
        }
        
        return answer;
    }
    
    public static void comb(String now) {
        if(now.length() > 5) {
            return;
        }
        
        if(!now.equals(""))  list.add(now);
        
        for(int i = 0; i < 5; i++) {
            comb(now + alphabet[i]);
        }
    }
}